src/VisualStudio/Core/Def/Implementation/DesignerAttribute/DesignerAttributeIncrementalAnalyzer.cs

```diff
@@ -48,147 +55,180 @@ internal partial class DesignerAttributeIncrementalAnalyzer : ForegroundThreadAf
+                    if (difference.TryGetValue(document.FilePath, out var result))
+                    {
+                        RegisterDesignerAttribute(document, result.DesignerAttributeArgument);
+                    }
                 }
+            }, _listener.BeginAsyncOperation("RegisterDesignerAttribute"));
+        }
 
-                uint itemId = vsDocument.GetItemId();
-                if (itemId == (uint)VSConstants.VSITEMID.Nil)
-                {
-                    // it is no longer part of the solution
-                    return;
-                }
+        private ImmutableDictionary<string, DesignerAttributeDocumentData> GetDifference(
+            ImmutableDictionary<string, DesignerAttributeDocumentData> oldFileToResult,
+            ImmutableDictionary<string, DesignerAttributeDocumentData> newFileToResult)
+        {
+            var difference = ImmutableDictionary.CreateBuilder<string, DesignerAttributeDocumentData>();
+
+            foreach (var newKvp in newFileToResult)
+            {
+                // 1) If this result is for a new document.  We always need to report it
+                // 2) If both the old and new data have this result, then report it if it is different.
+                var filePath = newKvp.Key;
+                var newResult = newKvp.Value;
 
-                if (ErrorHandler.Succeeded(vsDocument.Project.Hierarchy.GetProperty(itemId, (int)__VSHPROPID.VSHPROPID_ItemSubType, out var currentValue)))
+                if (!oldFileToResult.TryGetValue(filePath, out var oldResult) ||
+                    !newResult.Equals(oldResult))
                 {
-                    var currentStringValue = string.IsNullOrEmpty(currentValue as string) ? null : (string)currentValue;
-                    if (string.Equals(currentStringValue, designerAttributeArgument, StringComparison.OrdinalIgnoreCase))
-                    {
-                        // PERF: Avoid sending the message if the project system already has the current value.
-                        return;
-                    }
+                    difference.Add(filePath, newResult);
                 }
+            }
+
+            return difference.ToImmutable();
+        }
 
-                try
+        private void RegisterDesignerAttribute(Document document, string designerAttributeArgument)
+        {
+            var workspace = (VisualStudioWorkspaceImpl)document.Project.Solution.Workspace;
+
+            var vsDocument = workspace.GetHostDocument(document.Id);
+            if (vsDocument == null)
+            {
+                return;
+            }
+
+            uint itemId = vsDocument.GetItemId();
+            if (itemId == (uint)VSConstants.VSITEMID.Nil)
+            {
+                // it is no longer part of the solution
+                return;
+            }
+
+            if (ErrorHandler.Succeeded(vsDocument.Project.Hierarchy.GetProperty(itemId, (int)__VSHPROPID.VSHPROPID_ItemSubType, out var currentValue)))
```

Comment:
It's a shame that we have to do this for *every* document.  Would be nice if we could push that into register design view.

@davkean @srivatsn any ideas to improve this?